File: problem_2/haskell/problem2.hs
--fibinaci :: (Num a) => a -> a -> [a]
fibinaci a b
  | sum' > 4000000 = []
  | otherwise = sum' : fibinaci b sum'
  where sum' = a + b

problem2 = sum $ filter (even) (fibinaci 0 1)
problem2' = sum $ takeWhile (< 4000000) $ filter (even) $ scanl (\x acc -> x + acc) 0 [1..]

File: problem_2/rust/problem2.rs

fn main() {
   let iter = (1..).scan((0,1), |state, _| {
     *state = (state.1, state.0 + state.1);
     Some(*state)
   });

   let total = iter
               .take_while(|x| x.1 < 4000000)
               .filter(|x| x.1 % 2 == 0)
               .fold(0, |sum, i| sum + i.1);

   println!("{}", total);
}
